catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/transform/CsvTransform.java

```diff
@@ -0,0 +1,169 @@
+
+                    MetacardType metacardType = getMetacardType(properties,
+                            metacardTypes,
+                            attributeRegistry);
+
+                    Metacard metacard = new MetacardImpl(metacardType);
+
+                    properties.entrySet()
+                            .stream()
+                            .map(entry -> {
+                                String key = entry.getKey();
+                                Object val = entry.getValue();
+                                if (val == null) {
+                                    return null;
+                                }
+                                if (val instanceof List) {
+                                    List list = (List) val;
+                                    if (list.size() > 0 && !(list.get(0) instanceof Serializable)) {
+                                        return null;
+                                    }
+                                    return new AttributeImpl(key,
+                                            (List<Serializable>) new ArrayList<Serializable>((List) val));
+                                }
+                                if (!(val instanceof Serializable)) {
+                                    LOGGER.debug(
+                                            "Attribute value for [{}] is not serializable. Value: {}",
+                                            key,
+                                            val);
+                                    return null;
+                                }
+                                return new AttributeImpl(key, (Serializable) val);
+                            })
+                            .filter(Objects::nonNull)
+                            .forEach(metacard::setAttribute);
+
+                    return metacard;
+                })
+                .collect(Collectors.toList());
+    }
+
+    private MetacardType getMetacardType(Map<String, Object> mapMetacard,
+            List<MetacardType> metacardTypes, AttributeRegistry attributeRegistry) {
+        String metacardTypeName = (String) mapMetacard.getOrDefault("metacard-type",
+                "ddf.metacard");
+
+        MetacardType metacardType = metacardTypes.stream()
+                .filter(type -> type.getName()
+                        .equals(metacardTypeName))
+                .findFirst()
+                .orElse(new MetacardTypeImpl(metacardTypeName, ImmutableList.of()));
+
+        // Don't assume metacard type will contain all necessary attributes (eg, injected attributes)
+        Set<String> existingAttributeDescriptors = metacardType.getAttributeDescriptors()
+                .stream()
+                .map(AttributeDescriptor::getName)
+                .collect(Collectors.toSet());
+
+        Set<AttributeDescriptor> missingDescriptors = mapMetacard.keySet()
+                .stream()
+                .filter(key -> !existingAttributeDescriptors.contains(key))
```

Comment:
❓ I'm curious if it might be more efficient to call `Sets.difference(mapMetacard.keySet(), existingAttributeDescriptors)` before this stream and then use that result set and skip the filtering. Might be a wash.